File: mlproject/QMapArray.java
import java.util.Map;
import java.util.HashMap;

public class QMapArray implements QMap{

	private Map<Integer,double[]> qTable;
	int n;

	public QMapArray(int n){
		qTable = new HashMap<Integer,double[]>();
		this.n = n;
	}

	private static int getKey(int[] state){
		int key = 0;
		for(int v:state)
			key = key*10+v;
		return key;
	}

	public double[] get(int[] state){
		int key = getKey(state);
		if(!qTable.containsKey(key)){
			return null;
		}
		return qTable.get(key);
	}

	public void update(int[] state,double[] values){
		int key = getKey(state);
		qTable.put(key,values);
	}

	public void update(int[] lastState,Position lastAction,double qVal){
		double[] values = get(lastState);
		values[lastAction.i*n+lastAction.j] = qVal;
	}

}

File: python/qMapArray.py
#!/usr/bin/python
from qmap import QMap
class QMapArray(QMap):
	"""docstring for QMapArray"""
	def __init__(self,n):
		super(QMapArray, self).__init__()
		self.n = n
		self.qTable=dict()
	def getKey(self,state):
		key=0
		for v in state:
			key=key*10+v
		return key
	def get(self,state):
		key=self.getKey(state)
		if not key in self.qTable:
			self.qTable[key]=[0 for i in range(self.n*self.n)]
		return self.qTable[key]
	def update(self,lastState,lastAction,qVal):
		values=self.get(lastState)
		values[lastAction.i*self.n+lastAction.j]=qVal